packages/frontend/core/src/modules/explorer/views/tree/node.tsx

```diff
@@ -0,0 +1,417 @@
+  dropEffect?: ExplorerTreeNodeDropEffect;
+}) => {
+  const t = useI18n();
+  const cid = useId();
+  const context = useContext(ExplorerTreeContext);
+  const level = context?.level ?? 0;
+  // If no onClick or to is provided, clicking on the node will toggle the collapse state
+  const clickForCollapse = !onClick && !to && !disabled;
+  const [childCount, setChildCount] = useState(0);
+  const [renaming, setRenaming] = useState(defaultRenaming);
+  const [lastInGroup, setLastInGroup] = useState(false);
+  const rootRef = useRef<HTMLDivElement>(null);
+  const { dragRef, dragging, CustomDragPreview } = useDraggable<
+    AffineDNDData & { draggable: { __cid: string } }
+  >(
+    () => ({
+      data: { ...dndData?.draggable, __cid: cid },
+      dragPreviewPosition: 'pointer-outside',
+    }),
+    [cid, dndData]
+  );
+  const {
+    dropTargetRef,
+    treeInstruction,
+    draggedOverDraggable,
+    draggedOver,
+    draggedOverPosition,
+  } = useDropTarget<AffineDNDData & { draggable: { __cid: string } }>(
+    () => ({
+      data: dndData?.dropTarget,
+      treeInstruction: {
+        currentLevel: level,
+        indentPerLevel: 20,
+        mode: !collapsed
+          ? 'expanded'
+          : lastInGroup
+            ? 'last-in-group'
+            : 'standard',
+        block:
+          reorderable === false
+            ? ['reorder-above', 'reorder-below', 'reparent']
+            : [],
+      },
+      onDrop: data => {
+        if (
+          data.source.data.__cid === cid &&
+          data.treeInstruction?.type !== 'reparent'
+        ) {
+          // Do nothing if dropped on self
+          return;
+        }
+        onDrop?.(data);
+        if (data.treeInstruction?.type === 'make-child') {
+          setCollapsed(false);
+        }
+      },
+      canDrop,
+    }),
+    [
+      dndData?.dropTarget,
+      level,
+      collapsed,
+      lastInGroup,
+      reorderable,
+      canDrop,
+      cid,
+      onDrop,
+      setCollapsed,
+    ]
+  );
+  const isSelfDraggedOver = draggedOverDraggable?.data.__cid === cid;
+
+  useEffect(() => {
+    if (
+      draggedOver &&
+      treeInstruction?.type === 'make-child' &&
+      !isSelfDraggedOver
+    ) {
+      // auto expand when dragged over
+      const timeout = setTimeout(() => {
+        setCollapsed(false);
+      }, 1000);
+      return () => clearTimeout(timeout);
+    }
+    return;
+  }, [draggedOver, isSelfDraggedOver, setCollapsed, treeInstruction?.type]);
+
+  useEffect(() => {
+    if (rootRef.current) {
+      const parent = rootRef.current.parentElement;
+      if (parent) {
+        const updateLastInGroup = () => {
+          setLastInGroup(parent?.lastElementChild === rootRef.current);
+        };
+        updateLastInGroup();
+        const observer = new MutationObserver(updateLastInGroup);
+        observer.observe(parent, {
+          childList: true,
+        });
+        return () => observer.disconnect();
+      }
+    }
+    return;
+  }, []);
```

Comment:
How about record `children` on `context` and:
```tsx
const lastInGroup = context.children[context.children.length - 1].__cid === cid;
```

Reply:
There doesn't seem to be any good way to do this, and storing children list in the react context may cause a lot of re-rendering